README
To build this module you must first create a RELEASE_SITE
file. The RELEASE_SITE file should define the path to the
top of your EPICS tree at your site.

Example file looks like so:
===============================================================

#RELEASE Location of external products
# Run "gnumake clean uninstall install" in the application
# top directory each time this file is changed.

# =============================================================
# Define the top of the EPICS tree for your site.
# We will build some tools/scripts that allow us to
# change this easily when relocating software.
# ==============================================================
EPICS_SITE_TOP=/usr/local/lcls/epics

================================================================

Shantha Condamoor
19-Jun-2013:

The Caen ADC driver was modified:
    (1) To work with both 8-ch and 16-ch boards
	(2) 'old' and 'new' models
	(3) Several bug fixes from Spear (S.Allison) have been included in this release.
	
The driver is 'backward-compatible' - i.e., nothing needs to be done for a 16-ch board.
Call a different config function from st.cmd for 8-ch board

Example st.cmd that has one 16-ch board and one 8-ch board

# 16-channel board - same as before
caenV965Config( 0, 0x00A00000, 24, 90, 3, 0)

# 8-channel board - add "_8" to caenV965Config
caenV965Config_8( 1, 0x00C00000, 24, 91, 3, 0)
================================================================

Shantha Condamoor
4-June-2014:

Here is some insight into the relationship between the QDCRAW, threshold slope etc. and how
QDCRAW is calculated in the high gain and low gain modes:

When signals are at pedestal levels, QDCRAW and QDCRAWLO have about the same values.
ADC is in the high resolution (high gain) mode.

At this time, QDCRAW = (QDCRAWLO - lo threshold) * slope.

At modest signal levels (QDCRAW > 3840), QDCRAW and QDCRAWLO diverge;

At this time,

 QDCRAW = (8 x (QDCRAWHI - hi threshold)) * slope

since ADC is auto-ranging from the high resolution (high gain) channel to the lower gain channel.

It would definitely be useful to archive both HI and LO - 
even if only one of them is used in deriving QDCRAW at any time.

slope = QDCRAW.ASLO
High threshold  = QDCHITHRESH 
Low threshold = QDCLOTHRESH

For a sample display that makes all this mysterious calculations clear with labels,
look at $EDM/ws/pmtscreen_gadc.edl.

================================================================


RELEASE_NOTES
#==============================================================
#
#  Name: RELEASE_NOTES
#
#  Rem: This files describes release notes for
#       the Caen ADC V965 driver and device support.      
#
#  Auth: 01-Mar-2006  Author's Name      (DHT)
#  Rev:  dd-mmm-yyyy, Reviewer's Name    (USERNAME)
#--------------------------------------------------------------
#  Mod:
#        17-Jun-2013, S. Condamoor		 (SCONDAM):
#          caenADCV965-R2-2-4 - driver modified to work with both 8-ch and 16-ch boards.
#        20-Mar-2010, S. Hoobler         (SONYA):
#          release caenADCV965-R2-2-0:
#        30-Apr-2009, K. Luchini         (LUCHINI):
#          release caenADCV965-R2-0-0:
#        08-Oct-2007, K. Luchini         (LUCHINI):
#          release caenADCV965-R1-0-2
#          release caenADCV965-R1-0-1
#        17-Aug-2007, Doug Murray        (DRM):
#          release caenADCV965-R1-0
#
#==============================================================
#
R2.3.0             : Marcio Donadio
                     Modified on 25-Oct-2018

                                         Brought module to EPICS 3.15 and GIT workflow

caenADCV965-R2-2-6:  Shantha Condamoor 
                     Modified on 07-May-2014
					 
					 $(tse) macro added to the :QDCRAW record for BSA purposes.
					 
caenADCV965-R2-2-5:  Shantha Condamoor 
                     Modified on 02-May-2014
					 
					 Added LCLS-name specific ADC module and channel templates:
					 adc_module.template
					 adc_channel.template	
					 
caenADCV965-R2-2-4:  Shantha Condamoor 
                     Modified on 17-Jun-2013
					The Caen ADC driver was modified:
    					(1) To work with both 8-ch and 16-ch boards
						(2) 'old' and 'new' models
						(3) Several bug fixes from Spear (S.Allison) have been included in this release.
	
					The driver is 'backward-compatible' - i.e., nothing needs to be done for a 16-ch board.
					Call a different config function from st.cmd for 8-ch board

					Example st.cmd that has one 16-ch board and one 8-ch board:

					# 16-channel board - same as before
					caenV965Config( 0, 0x00A00000, 24, 90, 3, 0)

					# 8-channel board - add "_8" to caenV965Config
					caenV965Config_8( 1, 0x00C00000, 24, 91, 3, 0)				 
					 
caenADCV965-R2-2-3:  Change database Makefile to only use RULES 
                     not RULES.db in the local area

caenADCV965-R2-2-2: Release a tagged version of what is in MAIN_TRUNK.
06-Mar-2013

caenADCV965-R2-2-1:   Use $EPICS_BASE_VER macro in configure/RELEASE
01/19/2012            Include epicsTypes.h in drvV965.h
                      Build with base-R3-14-12

caenADCV965-R2-2-0:   Build with base-R3-14-8-2-lcls6
                      Build with rtems-4.9.4

caenADCV965-R2-1-0:   Build with base-R3-14-8-2-lcls5
                      Build with rtems-4.9.3

caenADCV965-R2-0-0: Build with base-R3-14-8-2-lcls3 (RTEMS-4.9.1)

caenADCV965-R1-0-1-1:
	modifications needed to bring the module to EPICS R3-14-9 standard

caenADCV965-R1-0-1: Build with base-R3-14-9-lcls1

caenADCV965-R1-0-1: Build with base-R3-14-8-2-lcls2

caenADCV965-R1-0:
===============================================================
Minor update - 8/17/2006 - Doug Murray
Added basic OSI support for RTEMS, and A32 addressing option
===============================================================


caenADCV965Lib/src/drvV965.h
// This file defines interfaces avaliable to the vxWorks shell and to record processing.

#include "dbScan.h"
#include "link.h" // from epics base:

#include <epicsEvent.h>
#include <epicsTypes.h>

/******************************************************\
|------------ Normal C interface points ---------------|
\******************************************************/

extern "C"
        {
        // Find a board if possible
        int caenV965Probe();

        // This configures up to 20 boards. For 16-ch boards
        int caenV965Config( int board, size_t base, int addrSpace, int vector, int level, int states);	
		
        // This configures up to 20 boards. For 8-ch boards
        int caenV965Config_8( int board, size_t base, int addrSpace, int vector, int level, int states);				

        // report function
        int drvCaenV965Report( int level);

        // Reset the state counter
        int drvCaenV965SetState( int board , int state);

        // Reset the state counter
        int drvCaenV965Wait( int board);
				
        }

// Config:
#define NUM_BOARDS (20)
#define CAEN_NUM_CHAN (16)

enum ADC_RANGE
        {
        ADC_HI = 0,
        ADC_LO = 1
        };


// These are declared in drvV965p.h
class drvCaenV965Registers;

// This is the low level device data structure.
// We create one of these per board.
class drvCaenV965Device
        {

        public:
                drvCaenV965Device();

                /* Initialize the board */

                static long init();
                static long report( int level);
                static int caenV965Config( int board, size_t base, int addrSpace, int vector, int level, int states);			
                static int caenV965Config_8( int board, size_t base, int addrSpace, int vector, int level, int states);					
				
                static void isr( void *pDev);
                static void atExit( void *arg);
                int readOutputBuffer();
                void show( int level);
                unsigned short getIped();
                void setIped( unsigned short);

                static int recordInit( DBLINK *pLink, dbCommon *pRec); // Do any per-record initialization
                int putValue( int signal, const char *parm, epicsInt32 value); // Return status
                static int putValue( DBLINK *pLink, epicsInt32 value); // Return status
                int getValue( int signal, const char *parm , epicsInt32 *value); // Return status
                static int getValue(DBLINK *pLink, epicsInt32  *value); // Return status
                static long getIOIntInfo( int cmd, DBLINK *pLink, IOSCANPVT *ppvt);
                void setState( int newState);
                long wait()
                        {

                        return epicsEventWait( wakeupCall);
                        }

                static drvCaenV965Device *getV965Handle( int board);

        private:

                drvCaenV965Registers *pBoard;
                                        // to end up invalid.

                // Process sync
                IOSCANPVT ioscanpvt;
                epicsEventId wakeupCall; // Set at the same time as scanIoRequest

                unsigned long event; // This is not the event counter on board
                                        // but rather is maintained by the ISR
                unsigned char int_vector;
                unsigned char int_level;
				int eight_ch_board;
				
                struct chanData
                        {
                        unsigned short data; // Same as range 0=>high 1=>low
                        unsigned short status; // 1=>under 2=over 0=normal
                        unsigned long event; // A copy of event when ISR last read this value
                        long gain; // Splice the high into the top of the low range - Only high channel
                                        // - 0 = no adjustment
                                        // 4096 is 1 count/count

                        unsigned short threshold; // The zero offset.
                        }chanData[CAEN_NUM_CHAN][2];

                int sampleState[CAEN_NUM_CHAN]; // When state is this then store the data
                int numStates; // Number of triggers per cycle
                int currentState; // Which trigger are we on
                static drvCaenV965Device *pDevice[NUM_BOARDS];     // We process all records on interrupt. All that did not process need
        };


caenADCV965Lib/src/devV965.cc
#include <longoutRecord.h>
#include <longinRecord.h>
#include <aiRecord.h>
#include <dbAccess.h>                           /* For S_db_badField constant macro */
#include <devSup.h>                             /* For device support entry table declarations */
#include <recGbl.h> 
#       ifndef EPICS_313
#       include <epicsExport.h> /* 3.14.4 Port */
#       endif

#include "drvV965.h"

/*
 * file:                devV965.cc
 * purpose:             EPICS Device Support for CAEN V965 VME Charge Integrating Dual Range ADC
 * created:             28-Oct-2005
 *                      Oak Ridge National Laboratory
 *
 * revision history:
 *   28-Oct-2005        David H Thompson        initial version
 *   17-Aug-2006        Doug Murray             updated
 */

///////////////////////////
//                       //
// ai device support     //
//                       //
///////////////////////////

struct AiDset
        {
        /*
         * analog input dset
         */
        long number;
        DEVSUPFUN dev_report;
        DEVSUPFUN init;
        DEVSUPFUN init_record;                  /*returns: (-1,0)=>(failure,success)*/
        DEVSUPFUN get_ioint_info;
        DEVSUPFUN read_ai;                      /*(0,2)=> success and convert,don't convert)*/
                                                /* if convert then raw value stored in rval */
        DEVSUPFUN special_linconv;
                                                /* Since this is c++ I can try this packagin scheme. */
        static long InitRecord( aiRecord *pRec);
        static long GetIOIntInfo( int cmd, aiRecord *pRec, IOSCANPVT *ppvt);
        static long ReadAi( aiRecord *pRec);
        }devCaenV965AI =
                {
                6,
                NULL,
                NULL,
                (DEVSUPFUN)AiDset::InitRecord,
                (DEVSUPFUN)AiDset::GetIOIntInfo,
                (DEVSUPFUN)AiDset::ReadAi,
                NULL
                };

//
//  device( ai, VME_IO,devCaenV965AI, "CAEN V965")
//
epicsExportAddress( dset, devCaenV965AI);

long AiDset::
InitRecord( aiRecord *pRec)
        {

        //pRec->linr=0;

        if( drvCaenV965Device::recordInit( &pRec->inp, (dbCommon *)pRec) != 0)
                {
                /* recGblRecordError(S_db_badField, (void *)pRec, "devCaenV965AI (init_record) Illegal INP field"); */
				/* scondam: 17-Jun-2013: Fix picked from Spear per S.Allison */
				recGblRecordError(S_db_badField, (void *)pRec, (char *)"devCaenV965AI (init_record) Illegal INP field");
                return S_db_badField;
                }

        // If we are reading the high register we use this scale
        if( pRec->inp.value.vmeio.parm[0] == 'H')
                pRec->eslo=200e-15;             // menuConvertSLOPE overrides
            else
                pRec->eslo=25e-15;              // menuConvertSLOPE overrides
        return 0;
        }

long AiDset::
GetIOIntInfo( int cmd, aiRecord *pRec, IOSCANPVT *ppvt)
        {

        return drvCaenV965Device::getIOIntInfo( cmd, &pRec->inp, ppvt);
        }

long AiDset::
ReadAi( aiRecord *pRec)
        {
        int rv;

        rv = drvCaenV965Device::getValue( &pRec->inp, &pRec->rval);

        // if( pRec->linr == menuConvertNO_CONVERSION)
        //        {
        //        pRec->val=pRec->rval;
        //        if( rv == OK)
        //              rv=2;
        //        }
        return rv; // dont convert 
        }

///////////////////////////
//                       //
// longin device support //
//                       //
///////////////////////////
struct LongInDset
        {
        long number;
        DEVSUPFUN dev_report;
        DEVSUPFUN init;
        DEVSUPFUN init_record;                  /* returns: (-1,0)=>(failure,success)*/
        DEVSUPFUN get_ioint_info;
        DEVSUPFUN read_longin;                  /* returns: (-1,0)=>(failure,success)*/
        static long InitRecord( longinRecord *pAI);
        static long GetIOIntInfo( int cmd, longinRecord *pAI, IOSCANPVT *ppvt);
        static long ReadLongin( longinRecord *pAI);
        }devCaenV965Longin =
                {
                5,
                NULL,
                NULL,
                (DEVSUPFUN)LongInDset::InitRecord,
                (DEVSUPFUN)LongInDset::GetIOIntInfo,
                (DEVSUPFUN)LongInDset::ReadLongin
                };

//
//  device( longin, VME_IO, devCaenV965Longin, "CAEN V965")
//
epicsExportAddress( dset, devCaenV965Longin);

long LongInDset::
InitRecord( longinRecord *pRec)
        {

        if( drvCaenV965Device::recordInit( &pRec->inp,(dbCommon *)pRec) != 0)
                {
                /* recGblRecordError( S_db_badField, (void *)pRec, "devCaenV965Longin (init_record) Illegal INP field"); */
				/* scondam: 17-Jun-2013: Fix picked from Spear per S.Allison */		
				recGblRecordError( S_db_badField, (void *)pRec, (char *)"devCaenV965Longin (init_record) Illegal INP field");		
                return S_db_badField;
                }
        return 0;
        }

long LongInDset::
GetIOIntInfo( int cmd, longinRecord *pRec, IOSCANPVT *ppvt)
        {

        return drvCaenV965Device::getIOIntInfo( cmd, &pRec->inp, ppvt);
        }

long LongInDset::
ReadLongin( longinRecord *pRec)
        {

        return drvCaenV965Device::getValue( &pRec->inp, &pRec->val);
        }

////////////////////////////
//                        //
// longout device support //
//                        //
////////////////////////////
struct LongOutDset
        {
        long number;
        DEVSUPFUN dev_report;
        DEVSUPFUN init;
        DEVSUPFUN init_record;                  /*returns: (-1,0)=>(failure,success) */
        DEVSUPFUN get_ioint_info;
        DEVSUPFUN write_longout;                /* (-1,0)=>(failure,success */
        static long InitRecord(longoutRecord *pRec);
        static long GetIOIntInfo(int cmd, longoutRecord *pRec, IOSCANPVT *ppvt);
        static long writeLongout(longoutRecord *pRec);
        }devCaenV965Longout =
                {
                5,
                NULL,
                NULL,
                (DEVSUPFUN)LongOutDset::InitRecord,
                (DEVSUPFUN)LongOutDset::GetIOIntInfo,
                (DEVSUPFUN)LongOutDset::writeLongout
                };

//
//  device( longout, VME_IO, devCaenV965Longout, "CAEN V965")
//
epicsExportAddress( dset, devCaenV965Longout);

long LongOutDset::
InitRecord( longoutRecord *pRec)
        {
        
        if( drvCaenV965Device::recordInit( &pRec->out, (dbCommon *)pRec) != 0)
                {
                /* recGblRecordError( S_db_badField, (void *)pRec, "devCaenV965Longout (init_record) Illegal OUT field"); */
				/* scondam: 17-Jun-2013: Fix picked from Spear per S.Allison */		
				recGblRecordError( S_db_badField, (void *)pRec, (char *)"devCaenV965Longout (init_record) Illegal OUT field");				
                return S_db_badField;
                }
        return 0;
        }

long LongOutDset::
GetIOIntInfo(int cmd, longoutRecord *pRec, IOSCANPVT *ppvt)
        {

        return drvCaenV965Device::getIOIntInfo( cmd, &pRec->out, ppvt);
        }

long LongOutDset::
writeLongout(longoutRecord *pRec)
        {

        return drvCaenV965Device::putValue( &pRec->out, pRec->val);
        }


caenADCV965Lib/src/drvV965p.h
// This is a driver for the caen v965 VME board.
// DH Thompson 10/28/2005
// DHT@ORNL.GOV

// This is a driver for the caen v965 VME board.
// DH Thompson 10/28/2005
// DHT@ORNL.GOV
#include "dbScan.h"
#include <libcpu/io.h>

#include "epicsMutex.h"

/*
 * file:                drvV965p.h
 * purpose:             Driver for 8-channel and 16-channel CAEN V965 VME Charge Integrating Dual Range ADC
 * created:             28-Oct-2005
 *                      Oak Ridge National Laboratory
 *
 * revision history:
 *   28-Oct-2005        David H Thompson        initial version
 *   17-Aug-2006        Doug Murray             updated
 */
		
class drvCaenV965Device;

class drvCaenV965Registers
        {
        // We will allow this class to access the registers directly where it makes sense.
        friend class drvCaenV965Device;

        public:
                static const int CAEN_NUM_SIGNALS = 16;
                /* static const int CAEN_NUM_SIGNALS = 8; //hmr for V965A				 */
                static const int CAEN_MODEL_NUMBER = 965;

                void show();
                inline int getModelNumber();
                inline long getSerialNumber();
                inline unsigned short getIped();
                inline void setIped( unsigned short val);
                inline int config( int vector, int level);
                inline void enableChannel( int chan, bool enable, int eight_ch_board);
                inline void enableLoChannel( int chan, bool enable, int eight_ch_board);
                inline void enableHiChannel( int chan, bool enable, int eight_ch_board);

        private:
                drvCaenV965Registers()
                        {
                        }; // No public constructor

                class CaenD16
                        {
                        public:
                                inline operator unsigned short();
                                inline unsigned short operator =( const unsigned short data);

                        private:
                                unsigned short loc;
                        };

                class CaenD32
                        {
                        public:
                                inline operator unsigned long();
                                inline unsigned long operator =( const unsigned long data);

                        private:
                                unsigned long loc;
                        };

                struct bitSet
                        {
                        CaenD16 set;
                        CaenD16 clear;
                        };

                /* struct ThresholdPair
                        {
                        CaenD16 high;
						// CaenD16 unused_0; //hmr for V965A 
                        CaenD16 low;					
						// CaenD16 unused_1;  //hmr for V965A
                        };
                */
				
                //
                // Data:
                // Registers: Reference page 33 of the manual
                //
                CaenD32 OutputBuffer[0x200];            // All are the same
                unsigned char unused_0[0x800];          // 0x800
                CaenD16 FirmwareRevision;               // 0x1000
                CaenD16 GeoAddress;                     // 0x1002
                CaenD16 MSCT_CBLT_Address;              // 0x1004
                bitSet bitSet1;                         // 0x1006
                CaenD16 InterruptLevel;                 // 0x100a
                CaenD16 InterruptVector;                // 0x100c
                CaenD16 StatusRegister1;                // 0x100e
                CaenD16 ControlRegister1;               // 0x1010
                CaenD16 ADERHigh;                       // 0x1012
                CaenD16 ADERLow;                        // 0x1014
                CaenD16 SingleShotReset;                // 0x1016
                unsigned char unused1[2];               // 0x1018
                CaenD16 MCST_CBLTCtrl;                  // 0x101a
                unsigned char unused2[4];               // 0x101c
                CaenD16 EventTriggerRegister;           // 0x1020
                CaenD16 StatusRegister2;                // 0x1022
                CaenD16 EventCounterL;                  // 0x1024
                CaenD16 EventCounterH;                  // 0x1026
                CaenD16 IncrementEvent;                 // 0x1028
                CaenD16 IncrementOffset;                // 0x102a
                CaenD16 LoadTestRegister;               // 0x102c
                CaenD16 FCLRWindow;                     // 0x102e
                unsigned char unused3[2];               // 0x1030
                bitSet bitSet2;                         // 0x1032
                CaenD16 WMemoryTestAddress;             // 0x1036
                CaenD16 MemoryTestWord_High;            // 0x1038
                CaenD16 MemoryTestWord_Low;             // 0x103a
                CaenD16 CrateSelect;                    // 0x103c
                CaenD16 TestEventWrite;                 // 0x103e
                CaenD16 EventCounterReset;              // 0x1040
                unsigned char unused4[0x1060 - 0x1042]; // 0x1042
                CaenD16 Iped;                           // 0x1060
                unsigned char unused5[2];               // 0x1062
                CaenD16 RTestAddress;                   // 0x1064
                unsigned char unused6[2];               // 0x1066
                CaenD16 SWComm;                         // 0x1068
                CaenD16 SlideConstant;                  // 0x106a
                unsigned char unused7[4];               // 0x106c
                CaenD16 AAD;                            // 0x1070
                CaenD16 BAD;                            // 0x1072
                unsigned char unused8[12];              // 0x1074
                /* ThresholdPair Thresholds[CAEN_NUM_SIGNALS]; // 0x1080-0x10bf				 */
				CaenD16  Thresholds[CAEN_NUM_SIGNALS*2]; // 0x1080-0x10bf	
                char unused9[0x8000 - 0x10c0];          // 0x10c0-0x8000
                CaenD16 ROM[0x4000];

                enum ROMIndex
                        {
                        OUI_MSB =       0,
                        OUI =           0x02a / sizeof( short),
                        OUI_LSB =       0x02e / sizeof( short),
                        Version =       0x032 / sizeof( short),
                        BoardIdMSB =    0x036 / sizeof( short),
                        BoardId =       0x03a / sizeof( short),
                        BoardIdLSB =    0x03e / sizeof( short),
                        MBRev =         0x04e / sizeof( short),
                        OBBRev =        0x052 / sizeof( short),
                        SerialMSB =     0xF02 / sizeof( short),
                        SerialLSB =     0xF06 / sizeof( short)
                        };

                enum OutputBufferWordType
                        {
                        OBT_header = 2<<24,
                        OBT_valid_datum = 0,
                        OBT_end_block = 4<<24,
                        OBT_not_valid_datum = 6<<24,
                        };

                enum OutputBufferBits
                        {
                        OBB_GEO_SHIFT = 1<<27,
                        OBB_GEO = 0x1f<<27,
                        OBT_mask = 7<<24,
                        // Header bits
                        OBB_CRATE_SHIFT = 1<<16,
                        OBB_CRATE = 0xff<<16,

                        OBB_CNT_SHIFT = 8,
                        OBB_CNT = 0x3f<<8,

                        // Datum bits
                        OBB_CHANNEL_SHIFT = 1<<17,										
                        OBB_CHANNEL = 0xf<<17,																				
                        OBB_RG = 1<<16,  // 4.5 is wrong?					
                        OBB_UN = 1<<13,
                        OBB_OV = 1<<12,
                        OBB_ADC = 0xfff<<0,
                        // EOB Bits
                        OBB_EVENT_COUNTER = 0xffffff<<0
                        };

                enum BitSet1Bits
                        {
                        BS1_BerrFlag = 1<<3,
                        BS1_SelAddr = 1<<4,
                        BS1_SoftReset = 1<<7
                        };

                enum BitSet2Bits
                        {
                        BS2_TestMem = 1<<0,
                        BS2_Offline = 1<<1,
                        BS2_ClearData = 1<<2,
                        BS2_OverRangeEn = 1<<3,

                        BS2_LowThresholdEn = 1<<4,
                        // Not used - 5
                        BS2_TestAc1 = 1<<6,
                        BS2_SlideEn = 1<<7,

                        // Reserved - 8
                        // not used - 9
                        // not used - 10
                        BS2_AutoIncr = 1<<11,

                        BS2_EmptyEnable = 1<<12,
                        BS2_SlideSubEn = 1<<13,
                        BS2_AllTrig = 1<<14
                        // Not used -15
                        };

                enum Status1Bits
                        {
                        ST1_Dready = 1<<0,
                        ST1_GlobalDready = 1<<1,
                        ST1_Busy = 1<<2,
                        ST1_GlobalBusy = 1<<3,
                        ST1_Purged = 1<<5,
                        ST1_EvRdy = 1<<8
                        };

                enum Status2Bits
                        {
                        ST2_BufferEmpty = 1<<1,
                        ST2_BufferFull = 1<<2,
                        // CSEL AND DSEL - get 965A vs 965
                        };
        };

inline drvCaenV965Registers::CaenD16::
operator unsigned short()
        {

        return in_be16( &loc);
        }

inline unsigned short drvCaenV965Registers::CaenD16::
operator =( const unsigned short data)
        {

        out_be16( &loc, data);
        return data;
        }

inline drvCaenV965Registers::CaenD32::
operator unsigned long()
        {

        return in_be32((volatile unsigned int *)&loc);
        }

inline unsigned long drvCaenV965Registers::CaenD32::
operator =( const unsigned long data)
        {

        out_be32((volatile unsigned int *)&loc, data);
        return data;
        }

inline unsigned short drvCaenV965Registers::
getIped()
        {

        return Iped;
        }

inline void drvCaenV965Registers::
setIped(unsigned short val)
        {

        Iped = val;
        }

inline void drvCaenV965Registers::
enableChannel(int chan, bool enable, int eight_ch_board)
        {
        unsigned short threshold;

        threshold = 0xff & Thresholds[(chan<<eight_ch_board)*2];	
        Thresholds[(chan<<eight_ch_board)*2] = threshold | (enable ? 0 : 0x100);
        threshold = 0xff & Thresholds[((chan<<eight_ch_board)*2)+(1<<eight_ch_board)];
        Thresholds[((chan<<eight_ch_board)*2)+(1<<eight_ch_board)] = threshold | (enable ? 0 : 0x100);
        }
		
inline void drvCaenV965Registers::
enableLoChannel(int chan, bool enable, int eight_ch_board)
        {
        unsigned short threshold;

        threshold = 0xff & Thresholds[((chan<<eight_ch_board)*2)+(1<<eight_ch_board)];			
        Thresholds[((chan<<eight_ch_board)*2)+(1<<eight_ch_board)] = threshold | (enable ? 0 : 0x100);
        }
		
inline void drvCaenV965Registers::
enableHiChannel(int chan, bool enable, int eight_ch_board)
        {
        unsigned short threshold;

        threshold = 0xff & Thresholds[(chan<<eight_ch_board)*2];			
        Thresholds[(chan<<eight_ch_board)*2] = threshold | (enable ? 0 : 0x100);
        }
			
inline int drvCaenV965Registers::
getModelNumber()
        {
        
        return((( ROM[BoardIdMSB] & 0xFF) << 16) | (( ROM[BoardId] & 0xFF) << 8) | ( ROM[BoardIdLSB] & 0xFF));
        }

inline long drvCaenV965Registers::
getSerialNumber()
        {
        long rv = 0;

        rv = ROM[SerialMSB] & 0xff;
		/* scondam: 17-Jun-2013: Fix picked from Spear per S.Allison */				
        rv = ( rv << 8) | (ROM[SerialLSB] & 0xff);
        return rv;
        }
